src/Microsoft.OData.Core/Json/JsonValueUtilsAsync.cs

```diff
@@ -0,0 +1,491 @@
+        internal static async Task WriteValueAsync(this TextWriter writer, decimal value)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            await writer.WriteAsync(value.ToString(CultureInfo.InvariantCulture)).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Write a DateTimeOffset value.
+        /// </summary>
+        /// <param name="writer">The text writer to write the output to.</param>
+        /// <param name="value">DateTimeOffset value to be written.</param>
+        /// <param name="dateTimeFormat">The format to write out the DateTime value in.</param>
+        [SuppressMessage("Microsoft.Globalization", "CA1303:Do not pass literals as localized parameters", MessageId = "Microsoft.OData.Json.JsonValueUtils.WriteQuotedAsync(System.IO.TextWriter,System.String)", Justification = "Constant defined by the JSON spec.")]
+        internal static async Task WriteValueAsync(this TextWriter writer, DateTimeOffset value, ODataJsonDateTimeFormat dateTimeFormat)
+        {
+            Debug.Assert(writer != null, "writer != null");
+
+            Int32 offsetMinutes = (Int32)value.Offset.TotalMinutes;
+
+            switch (dateTimeFormat)
+            {
+                case ODataJsonDateTimeFormat.ISO8601DateTime:
+                    {
+                        // Uses the same format as DateTime but with offset:
+                        // jsonDateTime= quotation-mark
+                        //  YYYY-MM-DDThh:mm:ss.sTZD
+                        //  [("+" / "-") offset]
+                        //  quotation-mark
+                        //
+                        // offset = 4DIGIT
+                        string textValue = XmlConvert.ToString(value);
+                        await writer.WriteQuotedAsync(textValue).ConfigureAwait(false);
+                    }
+
+                    break;
+
+                case ODataJsonDateTimeFormat.ODataDateTime:
+                    {
+                        // Uses the same format as DateTime but with offset:
+                        // jsonDateTime= quotation-mark
+                        //  "\/Date("
+                        //  ticks
+                        //  [("+" / "-") offset]
+                        //  ")\/"
+                        //  quotation-mark
+                        //
+                        // ticks = *DIGIT
+                        // offset = 4DIGIT
+                        string textValue = String.Format(
+                            CultureInfo.InvariantCulture,
+                            JsonConstants.ODataDateTimeOffsetFormat,
+                            DateTimeTicksToJsonTicks(value.Ticks),
+                            offsetMinutes >= 0 ? JsonConstants.ODataDateTimeOffsetPlusSign : string.Empty,
+                            offsetMinutes);
+                        await writer.WriteQuotedAsync(textValue).ConfigureAwait(false);
+                    }
+
+                    break;

```

Comment:
Is it possible to abstract that logic that generates the `textValue` in a method then re-use that method in both `WriteValueAsync` and `WriteValue`?

Reply:
@habbes Done